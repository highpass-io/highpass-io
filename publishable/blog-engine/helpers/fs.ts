import fs from "fs";

export function getFileContent(filePath: string): Promise<string> {
  return new Promise((resolve, reject) => {
    fs.readFile(filePath, (err, data) => {
      if (err) {
        reject(err);
      } else {
        resolve(String(data));
      }
    });
  });
}
