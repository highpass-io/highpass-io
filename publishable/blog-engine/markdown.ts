import * as Mdast from "mdast";
import unified from "unified";
import remark from "remark-parse";

import { getHighlightMarkup } from "./highlight";

type ToBeTyped = any;
type IncorrectlyTyped = any;

type ContentASTNodeType =
  | "blockquote"
  | "code"
  | "emphasis"
  | "fragment"
  | "h1"
  | "h2"
  | "h3"
  | "h4"
  | "h5"
  | "h6"
  | "html"
  | "inlineCode"
  | "image"
  | "link"
  | "listOrdered"
  | "listUnordered"
  | "listItem"
  | "paragraph"
  | "strong"
  | "text";

interface ContentASTNode {
  type: ContentASTNodeType;
  value?: ToBeTyped;
  children?: ContentASTNode[];
}

export interface ContentAST {
  type: "root";
  children: ContentASTNode[];
}

function getMdast(mdString: string): Mdast.Root {
  const processor = unified().use(remark);
  const mdast = processor.parse(mdString);

  const _mdast: any = mdast;
  const _mdastTyped: Mdast.Root = _mdast;
  return _mdastTyped;
}

function mdAstToContentAst(node: Mdast.Content): ContentASTNode {
  if (Array.isArray(node.children)) {
    let children: IncorrectlyTyped = node.children;
    let nodeType: ContentASTNodeType = "fragment";
    let value = undefined;

    if (node.type === "blockquote") {
      nodeType = "blockquote";
      const listItemChildren: IncorrectlyTyped = node.children;
      if (
        listItemChildren[0].type == "paragraph" &&
        listItemChildren.length === 1
      ) {
        children = node.children[0].children;
      }
    }

    if (node.type === "emphasis") {
      nodeType = "emphasis";
    }

    if (node.type === "heading") {
      if (node.depth === 1) nodeType = "h1";
      if (node.depth === 2) nodeType = "h2";
      if (node.depth === 3) nodeType = "h3";
      if (node.depth === 4) nodeType = "h4";
      if (node.depth === 5) nodeType = "h5";
      if (node.depth === 6) nodeType = "h6";
    }

    if (node.type === "inlineCode") {
      nodeType = "inlineCode";
    }

    if (node.type === "link") {
      nodeType = "link";
      value = {
        url: node.url,
        title: node.title,
      };
    }

    if (node.type === "list") {
      nodeType = node.ordered ? "listOrdered" : "listUnordered";
    }

    if (node.type === "listItem") {
      nodeType = "listItem";
      const listItemChildren: IncorrectlyTyped = node.children;
      if (
        listItemChildren[0].type == "paragraph" &&
        listItemChildren.length === 1
      ) {
        children = node.children[0].children;
      }
    }

    if (node.type === "paragraph") {
      nodeType = "paragraph";
    }

    if (node.type === "strong") {
      nodeType = "strong";
    }

    return {
      type: nodeType,
      value,
      children: children.map(mdAstToContentAst),
    };
  }

  let nodeType: ContentASTNodeType = "fragment";
  let value = node.value;

  if (node.type === "code") {
    nodeType = "code";
    value = {
      lang: node.lang,
      markup: getHighlightMarkup(node.value, node.lang || "text"),
    };
  }

  if (node.type === "html") {
    nodeType = "html";
    value = {
      markup: node.value,
    };
  }

  if (node.type === "inlineCode") {
    nodeType = "inlineCode";
  }

  if (node.type === "image") {
    nodeType = "image";
    value = {
      alt: node.alt,
      title: node.title,
      url: node.url,
    };
  }

  if (node.type === "text") {
    nodeType = "text";
  }

  return { type: nodeType, value: value };
}

export async function getContentAST(mdxString: string): Promise<ContentAST> {
  const mdast = getMdast(mdxString);

  return {
    type: "root",
    children: mdast.children.map(mdAstToContentAst),
  };
}

export function walkContentAST(
  ast: ContentAST | ContentASTNode,
  transformer: (n: ContentAST | ContentASTNode) => void
) {
  if (Array.isArray(ast.children)) {
    ast.children.forEach((a) => walkContentAST(a, transformer));
  }

  transformer(ast);
}
