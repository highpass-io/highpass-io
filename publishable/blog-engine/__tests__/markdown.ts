import path from "path";

import { getFileContent } from "../helpers/fs";
import { getHast } from "../markdown";

async function getExample(id: string): Promise<string> {
  const filePath = path.resolve(__dirname, `./fixtures/${id}.mdx`);
  return await getFileContent(filePath);
}

describe("markdown to html ast parser", () => {
  it("should parse headings", async () => {
    const markdown = "# Some heading";
    const ast = await getHast(markdown);
    expect(ast).toMatchSnapshot();
  });

  it("should parse headings + text", async () => {
    const markdown = "# Some heading\nSome text.";
    const ast = await getHast(markdown);
    expect(ast).toMatchSnapshot();
  });
});

describe("markdown to html ast parser full examples", () => {
  const exampleIds = ["simple", "simple-nested", "with-codeblock"];

  exampleIds.forEach((id) => {
    it(`should parse markdown ${id} example`, async () => {
      const markdown = await getExample(id);
      const ast = await getHast(markdown);
      expect(ast).toMatchSnapshot();
    });
  });
});
