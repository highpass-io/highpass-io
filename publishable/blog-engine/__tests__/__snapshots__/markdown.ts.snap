// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown to html ast parser full examples should parse markdown simple example 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 15,
              "line": 1,
              "offset": 14,
            },
            "start": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "text",
          "value": "Hello world!",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 15,
          "line": 1,
          "offset": 14,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {},
      "tagName": "h1",
      "type": "element",
    },
    Object {
      "type": "text",
      "value": "
",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 31,
              "line": 3,
              "offset": 46,
            },
            "start": Object {
              "column": 1,
              "line": 3,
              "offset": 16,
            },
          },
          "type": "text",
          "value": "This is just a simple example.",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 31,
          "line": 3,
          "offset": 46,
        },
        "start": Object {
          "column": 1,
          "line": 3,
          "offset": 16,
        },
      },
      "properties": Object {},
      "tagName": "p",
      "type": "element",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 1,
      "line": 4,
      "offset": 47,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`markdown to html ast parser full examples should parse markdown simple-nested example 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 9,
              "line": 1,
              "offset": 8,
            },
            "start": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "text",
          "value": "Hello ",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "world",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 16,
              "line": 1,
              "offset": 15,
            },
            "start": Object {
              "column": 9,
              "line": 1,
              "offset": 8,
            },
          },
          "properties": Object {},
          "tagName": "code",
          "type": "element",
        },
        Object {
          "position": Object {
            "end": Object {
              "column": 17,
              "line": 1,
              "offset": 16,
            },
            "start": Object {
              "column": 16,
              "line": 1,
              "offset": 15,
            },
          },
          "type": "text",
          "value": "!",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 17,
          "line": 1,
          "offset": 16,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {},
      "tagName": "h1",
      "type": "element",
    },
    Object {
      "type": "text",
      "value": "
",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 16,
              "line": 3,
              "offset": 33,
            },
            "start": Object {
              "column": 1,
              "line": 3,
              "offset": 18,
            },
          },
          "type": "text",
          "value": "This is just a ",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "simple",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 24,
              "line": 3,
              "offset": 41,
            },
            "start": Object {
              "column": 16,
              "line": 3,
              "offset": 33,
            },
          },
          "properties": Object {},
          "tagName": "code",
          "type": "element",
        },
        Object {
          "position": Object {
            "end": Object {
              "column": 33,
              "line": 3,
              "offset": 50,
            },
            "start": Object {
              "column": 24,
              "line": 3,
              "offset": 41,
            },
          },
          "type": "text",
          "value": " example.",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 33,
          "line": 3,
          "offset": 50,
        },
        "start": Object {
          "column": 1,
          "line": 3,
          "offset": 18,
        },
      },
      "properties": Object {},
      "tagName": "p",
      "type": "element",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 1,
      "line": 4,
      "offset": 51,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`markdown to html ast parser full examples should parse markdown with-codeblock example 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 15,
              "line": 1,
              "offset": 14,
            },
            "start": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "text",
          "value": "Hello world!",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 15,
          "line": 1,
          "offset": 14,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {},
      "tagName": "h1",
      "type": "element",
    },
    Object {
      "type": "text",
      "value": "
",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 23,
              "line": 3,
              "offset": 38,
            },
            "start": Object {
              "column": 1,
              "line": 3,
              "offset": 16,
            },
          },
          "type": "text",
          "value": "This has a code block!",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 23,
          "line": 3,
          "offset": 38,
        },
        "start": Object {
          "column": 1,
          "line": 3,
          "offset": 16,
        },
      },
      "properties": Object {},
      "tagName": "p",
      "type": "element",
    },
    Object {
      "type": "text",
      "value": "
",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "const foo = \\"bar\\";
",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 4,
              "line": 7,
              "offset": 68,
            },
            "start": Object {
              "column": 1,
              "line": 5,
              "offset": 40,
            },
          },
          "properties": Object {
            "className": Array [
              "language-js",
            ],
          },
          "tagName": "code",
          "type": "element",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 4,
          "line": 7,
          "offset": 68,
        },
        "start": Object {
          "column": 1,
          "line": 5,
          "offset": 40,
        },
      },
      "properties": Object {},
      "tagName": "pre",
      "type": "element",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 1,
      "line": 8,
      "offset": 69,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`markdown to html ast parser should parse headings + text 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 15,
              "line": 1,
              "offset": 14,
            },
            "start": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "text",
          "value": "Some heading",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 15,
          "line": 1,
          "offset": 14,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {},
      "tagName": "h1",
      "type": "element",
    },
    Object {
      "type": "text",
      "value": "
",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 11,
              "line": 2,
              "offset": 25,
            },
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 15,
            },
          },
          "type": "text",
          "value": "Some text.",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 11,
          "line": 2,
          "offset": 25,
        },
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 15,
        },
      },
      "properties": Object {},
      "tagName": "p",
      "type": "element",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 11,
      "line": 2,
      "offset": 25,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`markdown to html ast parser should parse headings 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 15,
              "line": 1,
              "offset": 14,
            },
            "start": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "text",
          "value": "Some heading",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 15,
          "line": 1,
          "offset": 14,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": Object {},
      "tagName": "h1",
      "type": "element",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 15,
      "line": 1,
      "offset": 14,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
